### ISSUE TITLE
Progenitor 0.6.0 position-dependant `allOf` `required` breakage

### ISSUE THREAD

**Issue by tormeh** [oxidecomputer/progenitor#738]
The following bug was introduced in the upgrade of Progenitor from 0.5.0 to 0.6.0, presumably in typify somewhere, although that's hard to say for certain.
With the following schema
```
{
  "components": {
    "schemas": {
      "Foo": {
        "allOf": [
          {
            "$ref": "#/components/schemas/Bar"
          },
          {
            "properties": {
              "baz": {
                "type": "boolean"
              }
            }
          }
        ],
        "required": [
          "baz"
        ]
      },
      "Bar": {
        "properties": {
          "type": {
            "type": "string"
          }
        }
      }
    }
  },
  "info": {
    "description": "",
    "title": "",
    "version": ""
  },
  "openapi": "3.0.1",
  "paths": {},
  "security": [
  ],
  "servers": [
  ],
  "tags": [
  ]
}
```
and this command: `RUST_BACKTRACE=1 cargo check`
I get
```
error: failed to run custom build command for `typify-require-bug v0.1.0 (/Users/i742018/Development/typify-require-bug)`
note: To improve backtraces for build dependencies, set the CARGO_PROFILE_DEV_BUILD_OVERRIDE_DEBUG=true environment variable to enable debug information generation.

Caused by:
  process didn't exit successfully: `/Users/i742018/Development/typify-require-bug/target/debug/build/typify-require-bug-f0075af65c28887c/build-script-build` (exit status: 101)
  --- stdout
  cargo:rerun-if-changed=./open_api_schema.json

  --- stderr
  thread 'main' panicked at /Users/i742018/.cargo/registry/src/index.crates.io-6f17d22bba15001f/typify-impl-0.0.16/src/merge.rs:268:70:
  called `Result::unwrap()` on an `Err` value: ()
  stack backtrace:
     0: rust_begin_unwind
               at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:645:5
     1: core::panicking::panic_fmt
               at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/panicking.rs:72:14
     2: core::result::unwrap_failed
               at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/result.rs:1653:5
     3: core::result::Result<T,E>::unwrap
     4: typify_impl::merge::merge_with_subschemas
     5: typify_impl::convert::<impl typify_impl::TypeSpace>::convert_schema_object
     6: typify_impl::convert::<impl typify_impl::TypeSpace>::convert_schema
     7: typify_impl::TypeSpace::convert_ref_type
     8: typify_impl::TypeSpace::add_ref_types_impl
     9: typify_impl::TypeSpace::add_ref_types
    10: progenitor_impl::Generator::generate_tokens
    11: build_script_build::main
    12: core::ops::function::FnOnce::call_once
  note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Reversing the allOf list fixes the issue. The following works:
```
{
  "components": {
    "schemas": {
      "Foo": {
        "allOf": [
          {
            "properties": {
              "baz": {
                "type": "boolean"
              }
            }
          },
          {
            "$ref": "#/components/schemas/Bar"
          }
        ],
        "required": [
          "baz"
        ]
      },
      "Bar": {
        "properties": {
          "type": {
            "type": "string"
          }
        }
      }
    }
  },
  "info": {
    "description": "",
    "title": "",
    "version": ""
  },
  "openapi": "3.0.1",
  "paths": {},
  "security": [
  ],
  "servers": [
  ],
  "tags": [
  ]
}
```

It seems like the `required` list can only reference properties in the first `allOf` entry. There is a bit more subtlety to this: Removing the `properties` of `Bar` also fixes the issue.

**Comment by ahl** [comment]
Thanks for filing this. Do you happen to know what this generated in 0.5.0? You're right that this is probably a typify bug; I'll take a look when I can get to it.

**Comment by tormeh** [comment]
Will take a look. In the meantime I made a PR for better typify errors: https://github.com/oxidecomputer/typify/pull/521

**Comment by tormeh** [comment]
It used to generate this:
```
pub use progenitor_client::{ByteStream, Error, ResponseValue};
#[allow(unused_imports)]
use progenitor_client::{encode_path, RequestBuilderExt};
#[allow(unused_imports)]
use reqwest::header::{HeaderMap, HeaderValue};
pub mod types {
    use serde::{Deserialize, Serialize};
    #[allow(unused_imports)]
    use std::convert::TryFrom;
    ///Bar
    ///
    /// <details><summary>JSON schema</summary>
    ///
    /// ```json
    /**{
  "properties": {
    "type": {
      "type": "string"
    }
  }
}*/
    /// ```
    /// </details>
    #[derive(Clone, Debug, Deserialize, Serialize)]
    pub struct Bar {
        #[serde(rename = "type", default, skip_serializing_if = "Option::is_none")]
        pub type_: Option<String>,
    }
    impl From<&Bar> for Bar {
        fn from(value: &Bar) -> Self {
            value.clone()
        }
    }
    ///Foo
    ///
    /// <details><summary>JSON schema</summary>
    ///
    /// ```json
    /**{
  "required": [
    "baz"
  ],
  "properties": {
    "baz": {
      "type": "boolean"
    },
    "type": {
      "type": "string"
    }
  }
}*/
    /// ```
    /// </details>
    #[derive(Clone, Debug, Deserialize, Serialize)]
    pub struct Foo {
        pub baz: bool,
        #[serde(rename = "type", default, skip_serializing_if = "Option::is_none")]
        pub type_: Option<String>,
    }
    impl From<&Foo> for Foo {
        fn from(value: &Foo) -> Self {
            value.clone()
        }
    }
}
#[derive(Clone, Debug)]
/**Client for



Version: */
pub struct Client {
    pub(crate) baseurl: String,
    pub(crate) client: reqwest::Client,
}
impl Client {
    /// Create a new client.
    ///
    /// `baseurl` is the base URL provided to the internal
    /// `reqwest::Client`, and should include a scheme and hostname,
    /// as well as port and a path stem if applicable.
    pub fn new(baseurl: &str) -> Self {
        #[cfg(not(target_arch = "wasm32"))]
        let client = {
            let dur = std::time::Duration::from_secs(15);
            reqwest::ClientBuilder::new().connect_timeout(dur).timeout(dur)
        };
        #[cfg(target_arch = "wasm32")]
        let client = reqwest::ClientBuilder::new();
        Self::new_with_client(baseurl, client.build().unwrap())
    }
    /// Construct a new client with an existing `reqwest::Client`,
    /// allowing more control over its configuration.
    ///
    /// `baseurl` is the base URL provided to the internal
    /// `reqwest::Client`, and should include a scheme and hostname,
    /// as well as port and a path stem if applicable.
    pub fn new_with_client(baseurl: &str, client: reqwest::Client) -> Self {
        Self {
            baseurl: baseurl.to_string(),
            client,
        }
    }
    /// Get the base URL to which requests are made.
    pub fn baseurl(&self) -> &String {
        &self.baseurl
    }
    /// Get the internal `reqwest::Client` used to make requests.
    pub fn client(&self) -> &reqwest::Client {
        &self.client
    }
    /// Get the version of this API.
    ///
    /// This string is pulled directly from the source OpenAPI
    /// document and may be in any format the API selects.
    pub fn api_version(&self) -> &'static str {
        ""
    }
}
impl Client {}
pub mod prelude {
    pub use super::Client;
}
```

**Comment by ahl** [comment]
I understand this better now. When introducing improved support for `not` constructions, I had to figure out how to handle the case where a particular property of an object was `required` to `not` be present. I handled this in a way that seemed a little loose, but expedient: remove that property from the properties map; f an object `required` a field not present in the `properties` map then we would infer that there was some logical conflict and decide that the schema was unsatisfiable.

This was a mistake (a bigger mistake than I expected). The schema you've cited is an example of why it's a mistake.

A more reasonable approach for a property that must `not` be present would be to set its schema to `false` i.e. unsatisfiable by any value. This also opens the door for schemas such as `{ "type": "object", "required": [ "foo" ] }` which is valid, but not something I've seen in the wild (`struct X { foo: serde_json::Value }`).

I tried this and it regresses some of the stuff I had worked on in that `allOf` merge stuff. Here's how I'd like to approach it:

1. Work I've been wanting to do around `oneOf` / `enum`s namely, getting subschemas into a canonical form and only then applying enum-categorization heuristics. This is beneficial independent of this regression in behavior, but would also make it easier not to regress the related merge work.

2. The resulting merged schemas with `not` + `required` could result in schemas more or less like this:

```json
{
  "type": "object",
  "properties": {
    "foo": false,
    "bar": { "type": "string" }
  }
}
```

We need to handle those `false` schema fields a little differently. Most simply we can ignore them. It would be pretty lousy to get a struct like this:

```rust
struct X {
    foo: Option<Never>,
    bar: Option<String>,
}

enum Never {}
```

We need the `struct`'s IR to represent these... somehow.

An even more sophisticated approach would include an `impl Deserialize` that **expressly** disallows those fields with `false` schemas (note that if `additionalProperties` is `false` we could just derive `Deserialize` since all unknown fields would be disallowed). This, however, is an area where we're already a bit loosey-goosey so it wouldn't hurt to defer that piece of it.
